### DISCUSSION TITLE
Support for SHACL Advanced Features

### DISCUSSION THREAD

**Discussion by Revnixcad** [dotnetrdf/dotnetrdf#559] [Q&A]
DotNetRDF already provides support for SHACL validation, but it currently lacks support for some of the advanced features of SHACL. These features include:

[Custom Targets](https://www.w3.org/TR/shacl-af/#dfn-custom-targets) - add flexibility to selecting the [focus nodes](https://www.w3.org/TR/shacl-af/#dfn-focus-node) of [shapes](https://www.w3.org/TR/shacl-af/#dfn-shape)
[Annotation Properties](https://www.w3.org/TR/shacl-af/#dfn-annotation-properties) - can create extra values in validation reports
[SHACL Functions](https://www.w3.org/TR/shacl-af/#dfn-shacl-functions) - encapsulate complex operations into reusable building blocks
[Node Expressions](https://www.w3.org/TR/shacl-af/#dfn-node-expressions) - describe how to derive sets of [nodes](https://www.w3.org/TR/shacl-af/#dfn-node)
[Expression Constraints](https://www.w3.org/TR/shacl-af/#dfn-expression-constraints) - enable constraint checks based on node expressions
[SHACL Rules](https://www.w3.org/TR/shacl-af/#dfn-shacl-rules) - enable deriving new [triples](https://www.w3.org/TR/shacl-af/#dfn-triple) from existing ones
Taken together or individually, these features greatly extend the application scenarios of SHACL, and SHACL-SPARQL in particular.

I'm curious if there are any plans to integrate these advanced features into the DotNetRDF library? Thank you in advance for your response.

**Comment by kal** [comment]
Hi @Revnixcad 

It would be nice to implement these features at some point in the future, but its not planned at the moment. My immediate focus is on reimplementing the SPARQL protocol functionality on .NET Core and working towards a semi-streaming implementation of the SPARQL query engine. If someone wanted to pick up doing some work on SHACL advanced features though I would be happy to support that!
